text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,527 @@
+implementation.  Synchronous instruments (Measure and its refinements)
+have three calling patterns (_Bound_, _Unbound_, and _Batch_) to
+capture measurements.  Asynchronous instruments (Observer and its
+refinements) use callbacks to capture measurements.
+
+All measurement APIs produce metric events consisting of [timestamp,
+instrument descriptor, label set, and numerical
+value](api-metrics.md#metric-event-format).  Synchronous instrument
+events additionally have [Context](api-context.md), describing
+properties of the associated trace and distributed correlation values.
+
+#### Terminology: Kinds of Aggregation
+
+_Aggregation_ refers to the technique used to summarize many
+measurements and/or observations into _some_ kind of summary of the
+data.  As detailed in the [metric SDK specification (TODO:
+WIP)](https://github.com/open-telemetry/opentelemetry-specification/pull/347/files?short_path=5b01bbf#diff-5b01bbf3430dde7fc5789b5919d03001),
+there are generally two relevant modes of aggregation:
+
+1. Within one collection interval, for one label set, the SDK's
+`Aggregator.Add()` interface method incorporates one new measurement
+value into the current aggregation value.  This happens at run time,
+therefore is referred to as _temporal aggregation_.  This mode applies
+only to Measure instruments.
+2. Within one collection interval, when combining label sets, the
+SDK's `Aggregator.Merge()` interface method incorporates two
+aggregation values into one aggregation value.  This is referred to as
+_spatial aggregation_.  This mode applies to both Measure and Observer
+instruments.
+
+As discussed below, we are especially interested in aggregating rate
+information, which sometimes requires that temporal and spatial
+aggregation be treated differently.
+
+### Last-value relationship
+
+Observer instruments have a well-defined _Last Value_ measured by the
+instrument, that can be useful in defining aggregations.  The Last
+Value of an Observer instrument is the value that was captured during
+the last-completed collection interval, and it is a useful
+relationship because it is defined without relation to collection
+interval timing.  The Last Value of an Observer is determined by the
+single most-recently completed collection interval--it is not
+necessary to consider prior collection intervals.  The Last Value of
+an Observer is undefined when it is not observed during a collection
+interval.
+
+To maintain this property, we impose a requirement: two or more
+`Observe()` calls with an identical LabelSet during a single Observer
+callback invocation are treated as duplicates of each other, where the
+last call to `Observe()` wins.
+
+Based on the Last Value relationship, we can ask and answer questions
+such as "what is the average last value of a metric at a point in
+time?".  Observer instruments define the Last Value relationship
+without referring to the collection interval and without ambiguity.
```

Comment:
Since "last value" is, by it's very nature, a single value for a given reporting interval, doesn't the "average" need to be defined over some set of collection intervals? And, if this is a "temporal" average (rather than a "spatial" average), doesn't there need to be some concept of a time window over which the average is computed?

Reply:
I think of average being well-defined for a single. interval. If there is no aggregation, there's a count of 1 and the average equals the single value. If there is spatial aggregation within the interval, the average is still well defined.

If there is temporal aggregation over multiple intervals, I think the average is still well defined. What we're not defining, however, is the average over a specific time range--the average is defined over a range of intervals.

I like the "last value" relationship defined here, since we can write queries that apply across machines, even though their collection intervals are not synchronized. The query is something like "read the last-value of metric X at time T across a set of machines and average the results". We are using the last value at the most-recent point in time as the input to a distributed spatial aggregation.

We could try to define an average across time and space using this data, but now some interpolation may be required, since the contributing points are spread across time and the query window is probably not aligned with the collection intervals of the target machines.

 See also https://github.com/open-telemetry/oteps/pull/96/files#diff-e822e6cd6cb430ee3f62ff7fbeaed0daR130

Reply:
I think it's good. I was just confused about the question when the answer seemed very obvious. :)